examples_tests/common/CommonAPI.h

```diff
@@ -293,6 +293,211 @@ class CommonAPI
+					extractedInfo.queueFamilyProps.compute = extractedInfo.queueFamilyProps.graphics;
+				}
+				
+				// If a unique Transfer is not found then use either Graphics or Compute for Transfer (prefer compute)
+				if(extractedInfo.queueFamilyProps.transfer.index == QueueFamilyProps::InvalidIndex)
+				{
+					if(extractedInfo.queueFamilyProps.compute.supportsTransfer)
+						extractedInfo.queueFamilyProps.transfer = extractedInfo.queueFamilyProps.compute;
+					else if(extractedInfo.queueFamilyProps.graphics.supportsTransfer)
+						extractedInfo.queueFamilyProps.transfer = extractedInfo.queueFamilyProps.graphics;
+				}
+			}
+
+			// Since our workload is not headless compute, a swapchain is mandatory
+			extractedInfo.isSwapChainSupported = gpu->isSwapchainSupported();
+
+			// Check if the surface is adequate
+			if(surface)
+			{
+				uint32_t surfaceFormatCount;
+				surface->getAvailableFormatsForPhysicalDevice(gpu, surfaceFormatCount, nullptr);
+				extractedInfo.availableSurfaceFormats = std::vector<video::ISurface::SFormat>(surfaceFormatCount);
+				surface->getAvailableFormatsForPhysicalDevice(gpu, surfaceFormatCount, extractedInfo.availableSurfaceFormats.data());
+
+				extractedInfo.availablePresentModes = surface->getAvailablePresentModesForPhysicalDevice(gpu);
+
+				extractedInfo.surfaceCapabilities = {};
+				if (surface->getSurfaceCapabilitiesForPhysicalDevice(gpu, extractedInfo.surfaceCapabilities))
+					extractedInfo.hasSurfaceCapabilities = true;
+			}
+		}
+
+		return extractedInfos;
+	}
+	
+	// TODO: also implement a function:findBestGPU
+	// Returns an index into gpus info vector
+	static uint32_t findSuitableGPU(const std::vector<GPUInfo>& extractedInfos, const bool graphicsQueueEnable)
+	{
+		uint32_t ret = ~0u;
+		for(uint32_t i = 0; i < extractedInfos.size(); ++i)
+		{
+			bool isGPUSuitable = false;
+			const auto& extractedInfo = extractedInfos[i];
+
+			if(graphicsQueueEnable)
+			{
+				if ((extractedInfo.queueFamilyProps.graphics.index != QueueFamilyProps::InvalidIndex) &&
+					(extractedInfo.queueFamilyProps.compute.index != QueueFamilyProps::InvalidIndex) &&
+					(extractedInfo.queueFamilyProps.transfer.index != QueueFamilyProps::InvalidIndex) &&
+					(extractedInfo.queueFamilyProps.present.index != QueueFamilyProps::InvalidIndex))
+					isGPUSuitable = true;
+			}
+			else
+			{
+				if ((extractedInfo.queueFamilyProps.compute.index != QueueFamilyProps::InvalidIndex) &&
+					(extractedInfo.queueFamilyProps.transfer.index != QueueFamilyProps::InvalidIndex))
+					isGPUSuitable = true;
+			}
+
+			if(extractedInfo.isSwapChainSupported == false)
+				isGPUSuitable = false;
+
+			if(extractedInfo.hasSurfaceCapabilities == false)
+				isGPUSuitable = false;
+
+			if(isGPUSuitable)
+			{
+				// find the first suitable GPU
+				ret = i;
+				break;
+			}
+		}
+
+		if(ret == ~0u)
+		{
+			assert(false);
+			ret = 0;
+		}
+
+		return ret;
+	}
 
 	template<uint32_t sc_image_count> //! input with window creation
```

Comment:
no idea why swapchain image count is a template argument

Reply:
It was here from the start and I don't know why it's the way it is.
@AnastaZIuk any ideas?

Reply:
probably because someone was banking on returning a `std::array<>` of swapchain images, but its moot because as you've correctly notices the actual number of swapchain images returned might be less and determined at runtime

So need to have this statically seared into CommonAPI

Reply:
Can I make it templated it on `MAX_SWAPCHAIN_IMAGE_COUNT` instead?
@devshgraphicsprogramming @Erfan-Ahmadi

Reply:
no need for any templates really